functions/profile.js

```diff
@@ -0,0 +1,107 @@
+const admin = require('./admin');
+const Promise = require('bluebird');
+const User = require("./user");
+
+// Trigger: /accounts/{userId}/githubToken
+// For legacy clients that only write githubToken (not profileNeedsUpdate)
+function enqueueUpdateLegacy(event) {
+  if (event.data.val()) {
+    return event.data.ref.parent.child('profileNeedsUpdate').set(true);
+  }
+}
+
+// Trigger: /accounts/{userId}/profileNeedsUpdate
+function enqueueUpdate(event) {
+  // Ignore profileNeedsUpdate=false
+  if (!event.data.val()) {
+    return;
+  }
+
+  // Create an entry in the queue
+  return writeQueueEntry(admin.database(), event.params.userId);
+}
+
+function writeQueueEntry(db, uid) {
+  const queueRef = db.ref(`/updateProfileQueue/${uid}`);
+  return queueRef.set({
+    createdAt: Math.floor(Date.now() / 1000),
+  });
+}
+
+// Trigger: /updateProfileQueue/{userId}
+function update(event) {
+  // Ignore deletes
+  if (!event.data.exists()) {
+    return;
+  }
+
+  const uid = event.params.userId;
+  const rootRef = admin.database().ref();
+
+  return User.findById(uid)
+    .then(user => {
+      // Fetch and save new profile data from GitHub
+      return user.updateProfile();
+    })
+    .then(() => {
+      // Atomically delete the queue entry and clear profileNeedsUpdate flag
+      const updates = {}
+      updates[`/updateProfileQueue/${uid}`] = null;
+      updates[`/accounts/${uid}/profileNeedsUpdate`] = false;
+      return rootRef.update(updates);
+    });
+}
+
+function cron() {
+  const db = admin.database();
+
+  function fixEnqueueUpdateLegacyFails() {
+    return db.ref('accounts').orderByChild('profileNeedsUpdate').equalTo(null).once('value')
+      .then(snapshot => {
+        var promises = [];
+        snapshot.forEach(data => {
+          if (data.val().githubToken) {
+            promises.push(db.ref(`/accounts/${data.key}/profileNeedsUpdate`).set(true));
+          }
+        });
+        return Promise.all(promises);
+      });
+  }
+
+  function fixEnqueueUpdateFails() {
+    return db.ref('accounts').orderByChild('profileNeedsUpdate').equalTo(true).once('value')
```

Comment:
when I first read this I wondered... what do we do with extra values in queue? then I realized writing them multiple times would have no negative side effect since writeQueueEntry only sets `createdAt` ... is that your intention?  I'm not sure I understand how `createdAt` and `updatedAt` are intended to work.  (Haven't reviewed updateProfile yet, which I assume is idempotent, but ideally we wouldn't re-trigger every task that is in-progress when the cron happens)